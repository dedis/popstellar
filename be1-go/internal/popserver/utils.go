package popserver

import (
	"github.com/rs/zerolog"
	"github.com/stretchr/testify/require"
	"go.dedis.ch/kyber/v3"
	"io"
	"popstellar/crypto"
	"popstellar/internal/popserver/repo"
	"popstellar/internal/popserver/types"
	"popstellar/message/query/method/message"
	"popstellar/network/socket"
	"popstellar/validation"
	"testing"
)

// FakeSocket is a fake implementation of a Socket
//
// - implements socket.Socket
type FakeSocket struct {
	socket.Socket

	ResultID    int
	Res         []message.Message
	MissingMsgs map[string][]message.Message
	Msg         []byte

	Err error

	// the Socket ID
	Id string
}

// Send implements socket.Socket
func (f *FakeSocket) Send(msg []byte) {
	f.Msg = msg
}

// SendResult implements socket.Socket
func (f *FakeSocket) SendResult(id int, res []message.Message, missingMsgs map[string][]message.Message) {
	f.ResultID = id
	f.Res = res
	f.MissingMsgs = missingMsgs
}

// SendError implements socket.Socket
func (f *FakeSocket) SendError(id *int, err error) {
	f.Err = err
}

func (f *FakeSocket) ID() string {
	return f.Id
}

func (f *FakeSocket) GetMessage() []byte {
	return f.Msg
}

func (f *FakeSocket) Type() socket.SocketType {
	return socket.ClientSocketType
}

func NewHandlerParameters(db repo.Repository) types.HandlerParameters {
	nolog := zerolog.New(io.Discard)
	schemaValidator, _ := validation.NewSchemaValidator()
	secret := crypto.Suite.Scalar().Pick(crypto.Suite.RandomStream())
	point := crypto.Suite.Point().Mul(secret, nil)
	return types.HandlerParameters{
		Log:                 nolog,
		Socket:              &FakeSocket{Id: "fakeID"},
		SchemaValidator:     *schemaValidator,
		DB:                  db,
		OwnerPubKey:         nil,
		ClientServerAddress: "ClientServerAddress",
		ServerServerAddress: "ServerServerAddress",
		ServerPubKey:        point,
		ServerSecretKey:     secret,
	}
}

func NewHandlerParametersWithOwnerAndServer(db repo.Repository, owner kyber.Point, server Keypair) types.HandlerParameters {
	nolog := zerolog.New(io.Discard)
	schemaValidator, _ := validation.NewSchemaValidator()

	return types.HandlerParameters{
		Log:                 nolog,
		Socket:              &FakeSocket{Id: "fakeID"},
		SchemaValidator:     *schemaValidator,
		DB:                  db,
		OwnerPubKey:         owner,
		ClientServerAddress: "ClientServerAddress",
		ServerServerAddress: "ServerServerAddress",
		ServerPubKey:        server.Public,
		ServerSecretKey:     server.Private,
	}
}

func NewHandlerParametersWithFakeSocket(db repo.Repository, s *FakeSocket) types.HandlerParameters {
	nolog := zerolog.New(io.Discard)
	schemaValidator, _ := validation.NewSchemaValidator()
	secret := crypto.Suite.Scalar().Pick(crypto.Suite.RandomStream())
	point := crypto.Suite.Point().Mul(secret, nil)

	return types.HandlerParameters{
		Log:                 nolog,
		Socket:              s,
		SchemaValidator:     *schemaValidator,
		DB:                  db,
		OwnerPubKey:         nil,
		ClientServerAddress: "ClientServerAddress",
		ServerServerAddress: "ServerServerAddress",
		ServerPubKey:        point,
		ServerSecretKey:     secret,
	}

}

type Keypair struct {
	Public     kyber.Point
	PublicBuf  []byte
	Private    kyber.Scalar
	PrivateBuf []byte
}

func GenerateKeyPair(t *testing.T) Keypair {
	secret := crypto.Suite.Scalar().Pick(crypto.Suite.RandomStream())
	point := crypto.Suite.Point().Mul(secret, nil)

	publicBuf, err := point.MarshalBinary()
	require.NoError(t, err)
	privateBuf, err := secret.MarshalBinary()

	return Keypair{point, publicBuf, secret, privateBuf}
}
