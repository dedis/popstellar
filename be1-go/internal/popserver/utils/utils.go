package utils

import (
	"github.com/rs/zerolog"
	"popstellar/validation"
	"sync"
)

var once sync.Once
var instance *utils

type utils struct {
	log             *zerolog.Logger
	schemaValidator *validation.SchemaValidator
}

func InitUtils(log *zerolog.Logger, schemaValidator *validation.SchemaValidator) {
	once.Do(func() {
		instance = &utils{
			log:             log,
			schemaValidator: schemaValidator,
		}
	})
}

func GetLogInstance() (*zerolog.Logger, bool) {
	if instance == nil || instance.log == nil {
		return nil, false
	}

	return instance.log, true
}

func GetSchemaValidatorInstance() (*validation.SchemaValidator, bool) {
	if instance == nil || instance.schemaValidator == nil {
		return nil, false
	}

	return instance.schemaValidator, true
}
