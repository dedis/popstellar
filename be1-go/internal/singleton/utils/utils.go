package utils

import (
	"github.com/rs/zerolog"
	"github.com/rs/zerolog/log"
	"popstellar/internal/errors"
	"popstellar/internal/validation"
	"sync"
)

var once sync.Once
var instance *utils

type utils struct {
	log             *zerolog.Logger
	schemaValidator *validation.SchemaValidator
}

func InitUtils(log *zerolog.Logger, schemaValidator *validation.SchemaValidator) {
	once.Do(func() {
		instance = &utils{
			log:             log,
			schemaValidator: schemaValidator,
		}
	})
}

func VerifyJSON(msg []byte, st validation.SchemaType) error {
	if instance == nil || instance.schemaValidator == nil {
		return errors.NewInternalServerError("schema validator was not instantiated")
	}

	return instance.schemaValidator.VerifyJSON(msg, st)
}

func LogInfo(msg string) {
	if instance == nil || instance.log == nil {
		return
	}

	instance.log.Info().Msg(msg)
}

func LogError(err error) {
	if instance == nil || instance.log == nil {
		return
	}

	log.Error().Msg(err.Error())
}
