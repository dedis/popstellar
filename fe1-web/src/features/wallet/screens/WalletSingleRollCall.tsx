import { CompositeScreenProps, useRoute } from '@react-navigation/core';
import { StackScreenProps } from '@react-navigation/stack';
import * as Clipboard from 'expo-clipboard';
import React, { useState } from 'react';
import { Modal, ScrollView, StyleSheet, Text, TextStyle, View } from 'react-native';
import { TouchableOpacity, TouchableWithoutFeedback } from 'react-native-gesture-handler';

import { PoPIcon, PoPTextButton, QRCode } from 'core/components';
import ModalHeader from 'core/components/ModalHeader';
import ScreenWrapper from 'core/components/ScreenWrapper';
import { AppParamList } from 'core/navigation/typing/AppParamList';
import { WalletParamList } from 'core/navigation/typing/WalletParamList';
import { ScannablePopToken } from 'core/objects/ScannablePopToken';
import { Color, Icon, ModalStyles, Spacing, Typography } from 'core/styles';
import STRINGS from 'resources/strings';

type NavigationProps = CompositeScreenProps<
  StackScreenProps<WalletParamList, typeof STRINGS.navigation_wallet_single_roll_call>,
  StackScreenProps<AppParamList, typeof STRINGS.navigation_app_lao>
>;

const styles = StyleSheet.create({
  publicKey: {
    marginVertical: Spacing.x2,
    color: Color.inactive,
    textAlign: 'center',
  } as TextStyle,
});

/**
 * Renders the screen for showing a single roll call in the wallet
 */
const WalletSingleRollCall = () => {
  const route = useRoute<NavigationProps['route']>();
  const { rollCallName } = route.params;

  return (
    <ScreenWrapper>
      <Text style={Typography.base}>{STRINGS.wallet_single_roll_call_description}</Text>
      <Text style={Typography.base}>{rollCallName}</Text>
    </ScreenWrapper>
  );
};

export default WalletSingleRollCall;

/**
 * Component shown in the top middle of the navigation bar. Makes sure
 * the name of a roll call is shown instead of some static text
 */
export const ViewSingleRollCallScreenHeader = () => {
  const route = useRoute<NavigationProps['route']>();
  const { rollCallName } = route.params;

  return (
    <Text style={Typography.topNavigationHeading} numberOfLines={1}>
      {rollCallName}
    </Text>
  );
};

/**
 * Component shown in the top right of the navigation bar. Allows the user to interact
 * show the qr code of their pop token.
 */
export const WalletSingleHeaderRight = () => {
  const route = useRoute<NavigationProps['route']>();
  const { rollCallTokenPublicKey } = route.params;

  const [modalVisible, setModalVisible] = useState(false);

  return (
    <>
      <TouchableOpacity onPress={() => setModalVisible(!modalVisible)}>
        <PoPIcon name="qrCode" color={Color.inactive} size={Icon.size} />
      </TouchableOpacity>

      <Modal
        transparent
        visible={modalVisible}
        onRequestClose={() => {
          setModalVisible(!modalVisible);
        }}>
        <TouchableWithoutFeedback
          containerStyle={ModalStyles.modalBackground}
          onPress={() => {
            setModalVisible(!modalVisible);
          }}
        />
        <ScrollView style={ModalStyles.modalContainer}>
          <ModalHeader onClose={() => setModalVisible(!modalVisible)}>
            {STRINGS.wallet_single_roll_call_pop_token}
          </ModalHeader>

          <View>
            <QRCode
              value={ScannablePopToken.encodePopToken({ pop_token: rollCallTokenPublicKey })}
            />
          </View>

          <Text style={[Typography.small, styles.publicKey]} selectable>
            {rollCallTokenPublicKey}
          </Text>

          <PoPTextButton onPress={() => Clipboard.setStringAsync(rollCallTokenPublicKey)}>
            {STRINGS.wallet_single_roll_call_copy_pop_token}
          </PoPTextButton>
        </ScrollView>
      </Modal>
    </>
  );
};
