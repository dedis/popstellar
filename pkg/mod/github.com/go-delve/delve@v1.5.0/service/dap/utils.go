package dap

// min returns the lowest-valued integer
// between the two passed into it.
func min(i, j int) int {
	if i < j {
		return i
	}
	return j
}
