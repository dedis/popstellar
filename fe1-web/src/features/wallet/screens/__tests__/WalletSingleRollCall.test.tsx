import { render } from '@testing-library/react-native';
import React from 'react';

import MockNavigator from '__tests__/components/MockNavigator';
import { mockPopToken } from '__tests__/utils';
import { mockRollCall } from 'features/rollCall/__tests__/utils';

import WalletSingleRollCall from '../WalletSingleRollCall';

describe('WalletSingleRollCall', () => {
  it('renders correctly', () => {
    const component = render(
      <MockNavigator
        component={WalletSingleRollCall}
        params={{
          rollCallId: mockRollCall.id.valueOf(),
          rollCallName: mockRollCall.name,
          rollCallTokenPublicKey: mockPopToken.publicKey.valueOf(),
        }}
      />,
    ).toJSON();
    expect(component).toMatchSnapshot();
  });
});

describe('ViewSingleRollCallScreenHeader', () => {
  it('renders correctly', () => {
    const component = render(
      <MockNavigator
        component={WalletSingleRollCall}
        params={{
          rollCallId: mockRollCall.id.valueOf(),
          rollCallName: mockRollCall.name,
          rollCallTokenPublicKey: mockPopToken.publicKey.valueOf(),
        }}
      />,
    ).toJSON();
    expect(component).toMatchSnapshot();
  });
});
