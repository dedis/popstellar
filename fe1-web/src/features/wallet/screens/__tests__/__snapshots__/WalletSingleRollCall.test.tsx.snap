// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewSingleRollCallScreenHeader renders correctly 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        [
          {
            "backgroundColor": "rgb(242, 242, 242)",
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <View
        collapsable={false}
        pointerEvents="box-none"
        style={
          {
            "zIndex": 1,
          }
        }
      >
        <View
          accessibilityElementsHidden={false}
          importantForAccessibility="auto"
          onLayout={[Function]}
          pointerEvents="box-none"
          style={null}
        >
          <View
            collapsable={false}
            pointerEvents="box-none"
            style={
              {
                "bottom": 0,
                "left": 0,
                "opacity": 1,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "zIndex": 0,
              }
            }
          >
            <View
              collapsable={false}
              style={
                {
                  "backgroundColor": "rgb(255, 255, 255)",
                  "borderBottomColor": "rgb(216, 216, 216)",
                  "flex": 1,
                  "shadowColor": "rgb(216, 216, 216)",
                  "shadowOffset": {
                    "height": 0.5,
                    "width": 0,
                  },
                  "shadowOpacity": 0.85,
                  "shadowRadius": 0,
                }
              }
            />
          </View>
          <View
            collapsable={false}
            pointerEvents="box-none"
            style={
              {
                "height": 44,
                "maxHeight": undefined,
                "minHeight": undefined,
                "opacity": undefined,
                "transform": undefined,
              }
            }
          >
            <View
              pointerEvents="none"
              style={
                {
                  "height": 0,
                }
              }
            />
            <View
              pointerEvents="box-none"
              style={
                {
                  "alignItems": "stretch",
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  {
                    "alignItems": "flex-start",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "justifyContent": "center",
                    "marginStart": 0,
                    "opacity": 1,
                  }
                }
              />
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  {
                    "justifyContent": "center",
                    "marginHorizontal": 16,
                    "maxWidth": 718,
                    "opacity": 1,
                  }
                }
              >
                <Text
                  accessibilityRole="header"
                  aria-level="1"
                  collapsable={false}
                  numberOfLines={1}
                  onLayout={[Function]}
                  style={
                    {
                      "color": "rgb(28, 28, 30)",
                      "fontSize": 17,
                      "fontWeight": "600",
                    }
                  }
                >
                  MockScreen
                </Text>
              </View>
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  {
                    "alignItems": "flex-end",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "justifyContent": "center",
                    "marginEnd": 0,
                    "opacity": 1,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
      <RNSScreenContainer
        onLayout={[Function]}
        style={
          {
            "flex": 1,
          }
        }
      >
        <RNSScreen
          activityState={2}
          collapsable={false}
          forwardedRef={[Function]}
          gestureResponseDistance={
            {
              "bottom": -1,
              "end": -1,
              "start": -1,
              "top": -1,
            }
          }
          pointerEvents="box-none"
          style={
            {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            collapsable={false}
            style={
              {
                "opacity": 1,
              }
            }
          />
          <View
            accessibilityElementsHidden={false}
            closing={false}
            gestureVelocityImpact={0.3}
            importantForAccessibility="auto"
            onClose={[Function]}
            onGestureBegin={[Function]}
            onGestureCanceled={[Function]}
            onGestureEnd={[Function]}
            onOpen={[Function]}
            onTransition={[Function]}
            pointerEvents="box-none"
            style={
              [
                {
                  "display": "flex",
                  "overflow": undefined,
                },
                {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
              ]
            }
            transitionSpec={
              {
                "close": {
                  "animation": "spring",
                  "config": {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
                "open": {
                  "animation": "spring",
                  "config": {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
              }
            }
          >
            <View
              collapsable={false}
              pointerEvents="box-none"
              style={
                {
                  "flex": 1,
                }
              }
            >
              <View
                collapsable={false}
                forwardedRef={[Function]}
                handlerTag={3}
                handlerType="PanGestureHandler"
                needsOffscreenAlphaCompositing={false}
                onGestureHandlerEvent={[Function]}
                onGestureHandlerStateChange={[Function]}
                style={
                  {
                    "flex": 1,
                    "transform": [
                      {
                        "translateX": 0,
                      },
                      {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  pointerEvents="box-none"
                  style={
                    [
                      {
                        "flex": 1,
                        "overflow": "hidden",
                      },
                      [
                        {
                          "backgroundColor": "rgb(242, 242, 242)",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <View
                    style={
                      {
                        "flex": 1,
                        "flexDirection": "column-reverse",
                      }
                    }
                  >
                    <View
                      style={
                        {
                          "flex": 1,
                        }
                      }
                    >
                      <View
                        style={
                          {
                            "flex": 1,
                          }
                        }
                      >
                        <RCTScrollView
                          collapsable={false}
                          handlerTag={4}
                          handlerType="NativeViewGestureHandler"
                          onGestureHandlerEvent={[Function]}
                          onGestureHandlerStateChange={[Function]}
                          style={
                            {
                              "padding": 16,
                            }
                          }
                          waitFor={
                            [
                              {
                                "current": null,
                              },
                            ]
                          }
                        >
                          <View>
                            <View>
                              <View
                                style={
                                  {
                                    "flexDirection": "row",
                                    "justifyContent": "center",
                                  }
                                }
                              >
                                <View
                                  style={
                                    {
                                      "position": "relative",
                                    }
                                  }
                                >
                                  <qrcode
                                    level="H"
                                    size={256}
                                    style={
                                      {
                                        "height": "auto",
                                        "maxWidth": "300px",
                                        "width": "100%",
                                      }
                                    }
                                    value="{"pop_token":"oKHk3AivbpNXk_SfFcHDaVHcCcY8IBfHE7auXJ7h4ms="}"
                                    viewBox="0 0 256 256"
                                  />
                                  <View
                                    style={
                                      {
                                        "alignItems": "center",
                                        "backgroundColor": "#3742fa",
                                        "borderRadius": 100,
                                        "bottom": "32.4%",
                                        "display": "flex",
                                        "justifyContent": "center",
                                        "left": "32.4%",
                                        "overflow": "hidden",
                                        "padding": "4",
                                        "position": "absolute",
                                        "right": "32.4%",
                                        "top": "32.4%",
                                      }
                                    }
                                  >
                                    <Text
                                      style={
                                        [
                                          {
                                            "fontSize": 16,
                                            "lineHeight": 20,
                                          },
                                          {
                                            "textAlign": "center",
                                          },
                                          {
                                            "color": "#fff",
                                          },
                                        ]
                                      }
                                    >
                                      Scan to get user token
                                    </Text>
                                  </View>
                                </View>
                              </View>
                            </View>
                            <Text
                              style={
                                [
                                  {
                                    "fontWeight": "bold",
                                  },
                                  {
                                    "textAlign": "center",
                                  },
                                  {
                                    "fontFamily": "Courier New",
                                  },
                                  {
                                    "marginTop": 32,
                                  },
                                ]
                              }
                            >
                              LoyalBorrow3877
                            </Text>
                            <Text
                              selectable={true}
                              style={
                                [
                                  {
                                    "fontSize": 16,
                                    "lineHeight": 20,
                                  },
                                  {
                                    "color": "#8E8E8E",
                                    "marginBottom": 16,
                                    "marginTop": 8,
                                    "textAlign": "center",
                                  },
                                  {
                                    "fontFamily": "Courier New",
                                  },
                                ]
                              }
                            >
                              oKHk3AivbpNXk_SfFcHDaVHcCcY8IBfHE7auXJ7h4ms=
                            </Text>
                            <View
                              accessible={true}
                              collapsable={false}
                              focusable={true}
                              onClick={[Function]}
                              onResponderGrant={[Function]}
                              onResponderMove={[Function]}
                              onResponderRelease={[Function]}
                              onResponderTerminate={[Function]}
                              onResponderTerminationRequest={[Function]}
                              onStartShouldSetResponder={[Function]}
                              style={
                                {
                                  "opacity": 1,
                                }
                              }
                            >
                              <View
                                style={
                                  [
                                    {
                                      "backgroundColor": "#3742fa",
                                      "borderColor": "#3742fa",
                                      "borderRadius": 8,
                                      "borderWidth": 1,
                                      "padding": 8,
                                    },
                                  ]
                                }
                              >
                                <Text
                                  style={
                                    [
                                      {
                                        "color": "#000",
                                        "fontSize": 20,
                                        "lineHeight": 26,
                                        "textAlign": "left",
                                      },
                                      {
                                        "textAlign": "center",
                                      },
                                      {
                                        "color": "#fff",
                                      },
                                    ]
                                  }
                                >
                                  Copy to Clipboard
                                </Text>
                              </View>
                            </View>
                          </View>
                        </RCTScrollView>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </RNSScreen>
      </RNSScreenContainer>
    </View>
  </RNCSafeAreaProvider>
</View>
`;

exports[`WalletSingleRollCall renders correctly 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        [
          {
            "backgroundColor": "rgb(242, 242, 242)",
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <View
        collapsable={false}
        pointerEvents="box-none"
        style={
          {
            "zIndex": 1,
          }
        }
      >
        <View
          accessibilityElementsHidden={false}
          importantForAccessibility="auto"
          onLayout={[Function]}
          pointerEvents="box-none"
          style={null}
        >
          <View
            collapsable={false}
            pointerEvents="box-none"
            style={
              {
                "bottom": 0,
                "left": 0,
                "opacity": 1,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "zIndex": 0,
              }
            }
          >
            <View
              collapsable={false}
              style={
                {
                  "backgroundColor": "rgb(255, 255, 255)",
                  "borderBottomColor": "rgb(216, 216, 216)",
                  "flex": 1,
                  "shadowColor": "rgb(216, 216, 216)",
                  "shadowOffset": {
                    "height": 0.5,
                    "width": 0,
                  },
                  "shadowOpacity": 0.85,
                  "shadowRadius": 0,
                }
              }
            />
          </View>
          <View
            collapsable={false}
            pointerEvents="box-none"
            style={
              {
                "height": 44,
                "maxHeight": undefined,
                "minHeight": undefined,
                "opacity": undefined,
                "transform": undefined,
              }
            }
          >
            <View
              pointerEvents="none"
              style={
                {
                  "height": 0,
                }
              }
            />
            <View
              pointerEvents="box-none"
              style={
                {
                  "alignItems": "stretch",
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  {
                    "alignItems": "flex-start",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "justifyContent": "center",
                    "marginStart": 0,
                    "opacity": 1,
                  }
                }
              />
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  {
                    "justifyContent": "center",
                    "marginHorizontal": 16,
                    "maxWidth": 718,
                    "opacity": 1,
                  }
                }
              >
                <Text
                  accessibilityRole="header"
                  aria-level="1"
                  collapsable={false}
                  numberOfLines={1}
                  onLayout={[Function]}
                  style={
                    {
                      "color": "rgb(28, 28, 30)",
                      "fontSize": 17,
                      "fontWeight": "600",
                    }
                  }
                >
                  MockScreen
                </Text>
              </View>
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  {
                    "alignItems": "flex-end",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "justifyContent": "center",
                    "marginEnd": 0,
                    "opacity": 1,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
      <RNSScreenContainer
        onLayout={[Function]}
        style={
          {
            "flex": 1,
          }
        }
      >
        <RNSScreen
          activityState={2}
          collapsable={false}
          forwardedRef={[Function]}
          gestureResponseDistance={
            {
              "bottom": -1,
              "end": -1,
              "start": -1,
              "top": -1,
            }
          }
          pointerEvents="box-none"
          style={
            {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            collapsable={false}
            style={
              {
                "opacity": 1,
              }
            }
          />
          <View
            accessibilityElementsHidden={false}
            closing={false}
            gestureVelocityImpact={0.3}
            importantForAccessibility="auto"
            onClose={[Function]}
            onGestureBegin={[Function]}
            onGestureCanceled={[Function]}
            onGestureEnd={[Function]}
            onOpen={[Function]}
            onTransition={[Function]}
            pointerEvents="box-none"
            style={
              [
                {
                  "display": "flex",
                  "overflow": undefined,
                },
                {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
              ]
            }
            transitionSpec={
              {
                "close": {
                  "animation": "spring",
                  "config": {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
                "open": {
                  "animation": "spring",
                  "config": {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
              }
            }
          >
            <View
              collapsable={false}
              pointerEvents="box-none"
              style={
                {
                  "flex": 1,
                }
              }
            >
              <View
                collapsable={false}
                forwardedRef={[Function]}
                handlerTag={1}
                handlerType="PanGestureHandler"
                needsOffscreenAlphaCompositing={false}
                onGestureHandlerEvent={[Function]}
                onGestureHandlerStateChange={[Function]}
                style={
                  {
                    "flex": 1,
                    "transform": [
                      {
                        "translateX": 0,
                      },
                      {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  pointerEvents="box-none"
                  style={
                    [
                      {
                        "flex": 1,
                        "overflow": "hidden",
                      },
                      [
                        {
                          "backgroundColor": "rgb(242, 242, 242)",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <View
                    style={
                      {
                        "flex": 1,
                        "flexDirection": "column-reverse",
                      }
                    }
                  >
                    <View
                      style={
                        {
                          "flex": 1,
                        }
                      }
                    >
                      <View
                        style={
                          {
                            "flex": 1,
                          }
                        }
                      >
                        <RCTScrollView
                          collapsable={false}
                          handlerTag={2}
                          handlerType="NativeViewGestureHandler"
                          onGestureHandlerEvent={[Function]}
                          onGestureHandlerStateChange={[Function]}
                          style={
                            {
                              "padding": 16,
                            }
                          }
                          waitFor={
                            [
                              {
                                "current": null,
                              },
                            ]
                          }
                        >
                          <View>
                            <View>
                              <View
                                style={
                                  {
                                    "flexDirection": "row",
                                    "justifyContent": "center",
                                  }
                                }
                              >
                                <View
                                  style={
                                    {
                                      "position": "relative",
                                    }
                                  }
                                >
                                  <qrcode
                                    level="H"
                                    size={256}
                                    style={
                                      {
                                        "height": "auto",
                                        "maxWidth": "300px",
                                        "width": "100%",
                                      }
                                    }
                                    value="{"pop_token":"oKHk3AivbpNXk_SfFcHDaVHcCcY8IBfHE7auXJ7h4ms="}"
                                    viewBox="0 0 256 256"
                                  />
                                  <View
                                    style={
                                      {
                                        "alignItems": "center",
                                        "backgroundColor": "#3742fa",
                                        "borderRadius": 100,
                                        "bottom": "32.4%",
                                        "display": "flex",
                                        "justifyContent": "center",
                                        "left": "32.4%",
                                        "overflow": "hidden",
                                        "padding": "4",
                                        "position": "absolute",
                                        "right": "32.4%",
                                        "top": "32.4%",
                                      }
                                    }
                                  >
                                    <Text
                                      style={
                                        [
                                          {
                                            "fontSize": 16,
                                            "lineHeight": 20,
                                          },
                                          {
                                            "textAlign": "center",
                                          },
                                          {
                                            "color": "#fff",
                                          },
                                        ]
                                      }
                                    >
                                      Scan to get user token
                                    </Text>
                                  </View>
                                </View>
                              </View>
                            </View>
                            <Text
                              style={
                                [
                                  {
                                    "fontWeight": "bold",
                                  },
                                  {
                                    "textAlign": "center",
                                  },
                                  {
                                    "fontFamily": "Courier New",
                                  },
                                  {
                                    "marginTop": 32,
                                  },
                                ]
                              }
                            >
                              LoyalBorrow3877
                            </Text>
                            <Text
                              selectable={true}
                              style={
                                [
                                  {
                                    "fontSize": 16,
                                    "lineHeight": 20,
                                  },
                                  {
                                    "color": "#8E8E8E",
                                    "marginBottom": 16,
                                    "marginTop": 8,
                                    "textAlign": "center",
                                  },
                                  {
                                    "fontFamily": "Courier New",
                                  },
                                ]
                              }
                            >
                              oKHk3AivbpNXk_SfFcHDaVHcCcY8IBfHE7auXJ7h4ms=
                            </Text>
                            <View
                              accessible={true}
                              collapsable={false}
                              focusable={true}
                              onClick={[Function]}
                              onResponderGrant={[Function]}
                              onResponderMove={[Function]}
                              onResponderRelease={[Function]}
                              onResponderTerminate={[Function]}
                              onResponderTerminationRequest={[Function]}
                              onStartShouldSetResponder={[Function]}
                              style={
                                {
                                  "opacity": 1,
                                }
                              }
                            >
                              <View
                                style={
                                  [
                                    {
                                      "backgroundColor": "#3742fa",
                                      "borderColor": "#3742fa",
                                      "borderRadius": 8,
                                      "borderWidth": 1,
                                      "padding": 8,
                                    },
                                  ]
                                }
                              >
                                <Text
                                  style={
                                    [
                                      {
                                        "color": "#000",
                                        "fontSize": 20,
                                        "lineHeight": 26,
                                        "textAlign": "left",
                                      },
                                      {
                                        "textAlign": "center",
                                      },
                                      {
                                        "color": "#fff",
                                      },
                                    ]
                                  }
                                >
                                  Copy to Clipboard
                                </Text>
                              </View>
                            </View>
                          </View>
                        </RCTScrollView>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </RNSScreen>
      </RNSScreenContainer>
    </View>
  </RNCSafeAreaProvider>
</View>
`;
